### PR TITLE
Add Logprobs unit test with a loose threshold

### PR DESCRIPTION
**Pull request by PrinsYin**
# Add Logprobs Unittest

## Motivation
We want to merge [#6318](https://github.com/sgl-project/sglang/pull/6318), so this PR adds a dedicated unittest suite to verify that the changes do not break **logprobs computation and interfaces**.  
The main goal is to catch indexing/corner-case issues and ensure stability/accuracy of logprobs against baseline data.

---

## Modifications
- Added `test/srt/test_logprobs.py`
  - Implements:
    - **`TestLogprobsDense`**: runs against `DEFAULT_SMALL_MODEL_NAME_FOR_TEST`
  - Configurable batch sizes, sample counts, and temperatures
  - Randomly samples records per run for coverage
  - Compares `input_top_logprobs` and `output_top_logprobs` against baselines
  - Validates both **max diff** and **mean diff** against thresholds
  - Writes per-config results into `$GITHUB_STEP_SUMMARY`

- Added to `run_suite.py` so it runs in CI
- Added retry logic for downloading Hugging Face baselines
- Enabled selective `return_logprob` in batches to check partial-request cases
- Cleaned up unsafe access and added tolerance-based failure reporting

---

## Baseline Data

WE ARE NOT TESTING MOE in THIS PR, the data below is just for the reference

- **Version**: `v0.5.1.post3`
- **Dense baseline**: [sglang_baseline_2000.pkl](https://huggingface.co/datasets/font-info/logprobs/blob/main/sglang_baseline_2000.pkl)
- **MoE baseline**: [sglang_baseline_moe.pkl](https://huggingface.co/datasets/font-info/logprobs/blob/main/sglang_baseline_moe.pkl)
- Generated ~1000 samples with `v0.5.1.post3`, uploaded to Hugging Face
- Each test run samples a few hundred records based on config
- Prompt length controlled to ~2000 tokens

---

## Tolerance Settings
**Dense (CUDA):**
- Max diff ≤ **1.5**
- Mean diff ≤ **0.1**

**Dense (ROCm):**
- Max diff ≤ **1.4**
- Mean diff ≤ **0.1**

**MoE (NVIDIA):**
- Max diff ≤ **10**
- Mean diff ≤ **0.3**

**MoE (ROCm):**
- Max diff ≤ **9.0**
- Mean diff ≤ **0.5**

---

## GitHub Summary Reporting
Each run appends a section like:

### CODE REVIEW

**Comment by yushengsu-thu** [comment]
Add AMD baseline and thresholds


- Dense:
Max diff ≤ 1.4
Mean diff ≤ 0.1 (per sample)

- MoE:
Max diff ≤ 9.0
Mean diff ≤ 0.5 (per sample)

**Comment by PrinsYin** [comment]
@narutolhy @zhaochenyang20 can you take a moment to review this? thanks!

File: test/srt/run_suite.py [hunk 1]
```diff
@@ -91,6 +91,7 @@ class TestFile:
         TestFile("test_no_chunked_prefill.py", 108),
         TestFile("test_no_overlap_scheduler.py", 234),
         TestFile("test_original_logprobs.py", 200),
+        TestFile("test_logprobs.py", 200),
```
Comment (narutolhy) [inline test/srt/run_suite.py]: This file sorts the test files alphabetically. test_logprobs.py can be added near the beginning of  `test_l`

File: test/srt/test_logprobs.py [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+    },
+    {
+        "batch_size": 20,
+        "num_samples": 1000,
+        "logprob_sample_ratio": 0.6,
+        "temperature": 1.0,
+    },
+]
+
+os.environ["RETURN_ORIGINAL_LOGPROB"] = "True"
```
Comment (narutolhy) [inline test/srt/test_logprobs.py]: Why do we need to turn on this switch? After turning it on, only the original logprobs will be returned. The different temperatures added in TEST_CONFIGS will not affect the returned logprobs.

File: test/srt/test_logprobs.py [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+from sglang.test.test_utils import (
+    DEFAULT_SMALL_MODEL_NAME_FOR_TEST,
+    write_github_step_summary,
+)
+
+if torch.version.hip is not None:
+    print("Running on AMD ROCm GPU")
+    MOE_INPUT_PKL_URL = "https://huggingface.co/datasets/yushengsu/logprobs/resolve/main/sglang_baseline_moe_v0.5.1.pkl"
+    MOE_TOLERANCE_MAX_DIFF = 9.0
+    MOE_TOLERANCE_MEAN_DIFF = 0.5
```
Comment (Hecate0821) [inline test/srt/test_logprobs.py]:
```suggestion
 
```
Configs never used

File: test/srt/test_logprobs.py [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+elif torch.version.cuda is not None:
+    print("Running on NVIDIA CUDA GPU")
+    MOE_INPUT_PKL_URL = "https://huggingface.co/datasets/font-info/logprobs/resolve/main/sglang_baseline_moe.pkl"
+    MOE_TOLERANCE_MAX_DIFF = 10
+    MOE_TOLERANCE_MEAN_DIFF = 0.3
```
Comment (Hecate0821) [inline test/srt/test_logprobs.py]:
```suggestion
 
```
Configs never used

File: test/srt/run_suite.py (hunk 1 above)
Comment (PrinsYin) [reply test/srt/run_suite.py]: thanks！

File: test/srt/test_logprobs.py (hunk 4 above)
Comment (PrinsYin) [reply test/srt/test_logprobs.py]: thanks！

File: test/srt/test_logprobs.py (hunk 2 above)
Comment (PrinsYin) [reply test/srt/test_logprobs.py]: I think we enable this so that logprobs stay independent of temperature and we can purely test consistency across different batch/ratio cases in this test? @zhaochenyang20

File: test/srt/test_logprobs.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+    DENSE_INPUT_PKL_URL = "https://huggingface.co/datasets/yushengsu/logprobs/resolve/main/sglang_baseline_v0.5.1.pkl"
+    DENSE_TOLERANCE_MAX_DIFF = 1.4
+    DENSE_TOLERANCE_MEAN_DIFF = 0.1
+elif torch.version.cuda is not None:
+    print("Running on NVIDIA CUDA GPU")
+    DENSE_INPUT_PKL_URL = "https://huggingface.co/datasets/font-info/logprobs/resolve/main/sglang_baseline_2000.pkl"
+    DENSE_TOLERANCE_MAX_DIFF = 1.5
+    DENSE_TOLERANCE_MEAN_DIFF = 0.1
+else:
+    print("No GPU backend (CPU only)")
```
Comment (zhaochenyang20) [inline test/srt/test_logprobs.py:31]: else raise error

File: test/srt/test_logprobs.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+        "num_samples": 1000,
+        "logprob_sample_ratio": 0.5,
+        "temperature": 2.0,
+    },
+    {
+        "batch_size": 20,
+        "num_samples": 1000,
+        "logprob_sample_ratio": 0.6,
+        "temperature": 1.0,
+    },
```
Comment (zhaochenyang20) [inline test/srt/test_logprobs.py]: shall we have a temperature `0`, `0.5` `1.0`, `2.0`? No need to have a `10.0`

File: test/srt/test_logprobs.py (hunk 2 above)
Comment (zhaochenyang20) [reply test/srt/test_logprobs.py]: Well. I think we do not need 4 group of tests. We only need one group of it. `num_samples 1000, ratio 0.5`.

File: test/srt/test_logprobs.py [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+    @classmethod
+    def setUpClass(cls):
+        """Set up the test class - initialize the engine once for all tests."""
+        print(f"Launching SGLang Engine with {DENSE_MODEL_NAME}...")
+        cls.engine = sgl.Engine(
+            model_path=DENSE_MODEL_NAME,
+            random_seed=42,
+            skip_tokenizer_init=True,
+            mem_fraction_static=0.6,
+            max_running_requests=1,
```
Comment (zhaochenyang20) [inline test/srt/test_logprobs.py]: `mem_fraction_static` should be higger.

why setting `max_running_requests`? We only need to set batch size when we are making the baseline. For the test, do not do this.

File: test/srt/test_logprobs.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+            )
+            for e1, e2 in zip(arrA, arrB):
+                if not e1 or not e2:
+                    continue
+                dmapA = {tid: lp for lp, tid, _ in e1}
+                dmapB = {tid: lp for lp, tid, _ in e2}
+                common = dmapA.keys() & dmapB.keys()
+                for tid in common:
+                    diffs.append(abs(dmapA[tid] - dmapB[tid]))
+        if not diffs:
```
Comment (zhaochenyang20) [inline test/srt/test_logprobs.py]: this if should have an else

File: test/srt/test_logprobs.py [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+                )
+                print(
+                    f"Will return logprobs for {logprob_count} samples (ratio: {config['logprob_sample_ratio']})"
+                )
+
+                all_max, all_mean = [], []
+                logprob_returned_count = 0
+
+                for i in range(0, len(test_records), config["batch_size"]):
+                    batch = test_records[i : i + config["batch_size"]]
```
Comment (zhaochenyang20) [inline test/srt/test_logprobs.py]: why do we need to have a batch size. Could we send all the 1000 requests all at once?

File: test/srt/test_logprobs.py [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+                            input_top_logprobs = (
+                                meta_info.get("input_top_logprobs")
+                                if meta_info
+                                else None
+                            )
+
+                            self.assertIsNotNone(
+                                input_top_logprobs,
+                                f"return_logprob enabled on this sample, but input_top_logprobs is None (length: {len(input_top_logprobs) if input_top_logprobs is not None else 'N/A'})",
+                            )
```
Comment (zhaochenyang20) [inline test/srt/test_logprobs.py]: also assert these we do not want log probs actaully do not return anything.

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

File: test/srt/test_logprobs.py (hunk 7 above)
Comment (PrinsYin) [reply test/srt/test_logprobs.py]: oops, my bad

File: test/srt/test_logprobs.py (hunk 10 above)
Comment (PrinsYin) [reply test/srt/test_logprobs.py]: It is asserted on line 236

File: test/srt/test_logprobs.py (hunk 9 above)
Comment (PrinsYin) [reply test/srt/test_logprobs.py]: ofc!

File: test/srt/test_logprobs.py (hunk 2 above)
Comment (narutolhy) [reply test/srt/test_logprobs.py]: I feel that if this is added, there is no need to compare the effects of different temperatures, because different temperatures do not affect the return value.

File: test/srt/test_logprobs.py [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+TOP_K = 20
+MAX_RETRIES = 3
+RETRY_DELAY = 2
+
+# Test configurations
+TEST_CONFIG = {
+    "num_samples": 1000,
+    "logprob_sample_ratio": 0.5,
+    "temperature": 1.0,
+}
```
Comment (zhaochenyang20) [inline test/srt/test_logprobs.py]: No need to use a dict for this. Just as these:

```python
TOP_K = 20
MAX_RETRIES = 3
RETRY_DELAY = 2
```

**Review by zhaochenyang20 — APPROVED** [review]

File: test/srt/test_logprobs.py [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+
+            except Exception as e:
+                print(f"Attempt {attempt + 1}/{MAX_RETRIES} failed: {e}")
+                if attempt == MAX_RETRIES - 1:
+                    raise Exception(
+                        f"Failed to load data after {MAX_RETRIES} attempts: {e}"
+                    )
+                time.sleep(RETRY_DELAY)
+
+    def compare_meta(self, metaA, metaB):
```
Comment (zhaochenyang20) [inline test/srt/test_logprobs.py]: do not name it as A or B. Use something like `baseline`.

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

**Comment by zhaochenyang20** [comment]
Tested on H200 for 5 times:

```python
max Δ=0.624243
mean Δ=0.0032411
logprobs returned for 500 samples (expected: 500)
.
max Δ=0.624243
mean Δ=0.00346705
logprobs returned for 500 samples (expected: 500)
.
max Δ=0.624243
mean Δ=0.00395164
logprobs returned for 500 samples (expected: 500)

max Δ=0.624243
mean Δ=0.00399972
logprobs returned for 500 samples (expected: 500)

  loop = asyncio.get_event_loop()
max Δ=0.499887
mean Δ=0.00326239
logprobs returned for 500 samples (expected: 500)
```

**Review by merrymercy — APPROVED** [review]

**Review by zhaochenyang20 — APPROVED** [review]
